### PR TITLE
Allow a Collection of Proto-Fragments or a Collection of Fragment-Instances to be detached in single call

### PR DESCRIPTION
**Pull request by aang114**
Refactored `fragments.detach()` and `protos.detach()` to accept a list of Proto-Fragments and a list of Fragment-Instances (of a single non-copyable FD) respectively.

By calling any one of the two extrisnics - it will **ultimately cause an event will be emitted** which will **contain the merkle-root of the list of protos/FIs**

### CODE REVIEW

File: pallets/detach/src/lib.rs [hunk 1]
```diff
@@ -486,7 +503,8 @@ pub mod pallet {
 					// We can still have multiple transactions compete for the same "spot",
 					// and the one with higher priority will replace other one in the pool.
 					.and_provides((
-						data.hash.clone(),
+						data.hashes.clone(), // TODO Review - Isn't it redundant to have `data.hashes` as part of the "provides" tag? Isn't `data.merkle_root` enough?
```
Comment (sinkingsugar) [inline pallets/detach/src/lib.rs]: I think it's redundant for the reason you mention yes.

File: pallets/fragments/src/lib.rs [hunk 2]
```diff
@@ -1383,45 +1384,63 @@ pub mod pallet {
...
 			ensure!(
-				owned_instances.contains(&(Compact(edition_id), Compact(copy_id))),
+				(definition_permissions & FragmentPerms::COPY) != FragmentPerms::COPY,
 				Error::<T>::NoPermission
 			);
 
-			// REVIEW - no go if will expire this block
-			let instance = <Fragments<T>>::get((definition_hash, edition_id, copy_id))
-				.ok_or(Error::<T>::NotFound)?;
-			if let Some(expiring_at) = instance.expiring_at {
```
Comment (sinkingsugar) [inline pallets/fragments/src/lib.rs:1416]: I think for now any item that can expire should not be detached at all

**Review by sinkingsugar — CHANGES_REQUESTED** [review]
Great! I think we need more tests though.

File: pallets/fragments/src/lib.rs [hunk 3]
```diff
@@ -1383,45 +1384,63 @@ pub mod pallet {
...
 			target_account: BoundedVec<u8, T::DetachAccountLimit>, // an eth address or so
 		) -> DispatchResult {
-
 			let who = ensure_signed(origin)?;
 
+			ensure!(!edition_ids.is_empty(), Error::<T>::InstancesToDetachIsEmpty);
 			ensure!(!target_account.is_empty(), Error::<T>::DetachAccountIsEmpty);
 
-			let current_block_number = <frame_system::Pallet<T>>::block_number();
+			// let current_block_number = <frame_system::Pallet<T>>::block_number();
```
Comment (sinkingsugar) [inline pallets/fragments/src/lib.rs]: We don't use this?

File: pallets/fragments/src/lib.rs [hunk 4]
```diff
@@ -1383,45 +1384,63 @@ pub mod pallet {
...
+
+			let definition_permissions =
+				<Definitions<T>>::get(definition_hash).ok_or(Error::<T>::NotFound)?.permissions;
 
-			// owner must own instance
-			let owned_instances = <Inventory<T>>::get(who, definition_hash).ok_or(Error::<T>::NoPermission)?;
 			ensure!(
-				owned_instances.contains(&(Compact(edition_id), Compact(copy_id))),
+				(definition_permissions & FragmentPerms::COPY) != FragmentPerms::COPY,
```
Comment (sinkingsugar) [inline pallets/fragments/src/lib.rs]: btw also COPIABLE items cannot be detached I guess for now

File: pallets/fragments/src/lib.rs (hunk 2 above)
Comment (aang114) [reply pallets/fragments/src/lib.rs:1416]: But if a Fragment Definition is not copyable, it cannot expire. That's why I don't check

File: pallets/fragments/src/lib.rs (hunk 3 above)
Comment (aang114) [reply pallets/fragments/src/lib.rs]: No we don't. Since we don't need to check for expiry since **if a Fragment Definition is not copyable, it cannot expire**

File: pallets/fragments/src/lib.rs (hunk 4 above)
Comment (aang114) [reply pallets/fragments/src/lib.rs]: Yes. That's why I check if the Fragment Definition is copyable - rather than checking if the item is copyable

Comment (aang114) [reply pallets/fragments/src/lib.rs]: > btw also COPIABLE items cannot be detached I guess for now

Hmm but some copyable items can be `give()`en and its copy permission removed 🤔. That's true.

**So should I also accept Copy ID(s) as a parameter? ** Cos in the spec you previously mentioned, you only wanted to accept `edition_ids` @sinkingsugar

Comment (sinkingsugar) [reply pallets/fragments/src/lib.rs]: Did we clarify this on discord @aang114 ?

File: pallets/detach/src/lib.rs [hunk 5]
```diff
@@ -486,7 +543,7 @@ pub mod pallet {
 					// We can still have multiple transactions compete for the same "spot",
 					// and the one with higher priority will replace other one in the pool.
 					.and_provides((
-						data.hash.clone(),
+						data.merkle_root, // TODO Review - what if by co-incidence two collections that contain different things have the same merkle-root?
```
Comment (alebaffa) [inline pallets/detach/src/lib.rs]: With a 256 bit hash function, a merkle root collision is likely to be probabilistically impossible. In general, from my readings, with a hash function >= 64 bit you can be relatively safe.

Comment (sinkingsugar) [reply pallets/detach/src/lib.rs]: To answer his comment.
Stars would collide and we would be sucked by a black hole if that happens.

Comment (sinkingsugar) [reply pallets/detach/src/lib.rs]: It's blake2 no?

**Review by sinkingsugar — CHANGES_REQUESTED** [review]

File: pallets/detach/src/uomo_contento.rs [hunk 6]
```diff
@@ -0,0 +1,140 @@
+//! Everything in this crate was copied from: https://crates.io/crates/beefy-merkle-tree/10.0.0
```
[the remark this answers is not in the document]
Comment (alebaffa) [reply pallets/detach/src/uomo_contento.rs:1]: oh thank you, seriously. I was about to write the same comment.

[the remark this answers is not in the document]
Comment (alebaffa) [reply pallets/detach/src/uomo_contento.rs:1]: A note: always keep in mind that the code you write today will likely be seen/modified/inspected by someone else months/years from now. And it may also happen that you will not be here anymore to help that person. So, meaningful naming is important in a code base, as a knowledge sharing practice.

**Review by alebaffa — APPROVED** [review]

File: pallets/detach/src/uomo_contento.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (aang114) [reply pallets/detach/src/uomo_contento.rs:1]: > Please give a meaningful name to this file.

I named it this because I wanted to emphasize that this is only a temporary arrangement (hopefully)

File: pallets/detach/src/lib.rs (hunk 5 above)
Comment (aang114) [reply pallets/detach/src/lib.rs]: > It's blake2 no?

the merkle root is calculated using keccak256

**Comment by aang114** [comment]
Hi @sinkingsugar, please approve this PR (if okay) so that I can merge it. Thanks

File: pallets/detach/src/lib.rs (hunk 5 above)
Comment (sinkingsugar) [reply pallets/detach/src/lib.rs]: that's ok too!

File: pallets/detach/src/copy_pasted_beefy_merkle_tree.rs [hunk 7]
```diff
@@ -0,0 +1,144 @@
+//! Everything in this crate was copied from: https://crates.io/crates/beefy-merkle-tree/10.0.0 (which is also used in Substrate)
```
Comment (sinkingsugar) [inline pallets/detach/src/copy_pasted_beefy_merkle_tree.rs:1]: less silly name please.
Can you simply seriously call it `merkle_tree.rs`? You explained the rest in comments...

**Review by sinkingsugar — CHANGES_REQUESTED** [review]

File: pallets/detach/src/copy_pasted_beefy_merkle_tree.rs (hunk 7 above)
Comment (aang114) [reply pallets/detach/src/copy_pasted_beefy_merkle_tree.rs]: done

File: pallets/detach/src/lib.rs [hunk 8]
```diff
@@ -85,22 +87,49 @@ use frame_system::offchain::{
...
+pub enum DetachCollection {
+	/// List of Proto-Fragments (identified by its Hash)
+	Protos(Vec<Hash256>),
+	/// List of Fragment Instances (identified as a tuple of its Fragment Definition Hash, its Edition ID and its Copy ID)
+	Instances(Vec<(Hash128, Compact<InstanceUnit>, Compact<InstanceUnit>)>),
+}
+impl DetachCollection {
+
+	/// Get the Collection type
+	fn get_type(&self) -> Vec<u8> {
```
Comment (sinkingsugar) [inline pallets/detach/src/lib.rs]: What's the point of this?

File: pallets/detach/src/lib.rs [hunk 9]
```diff
@@ -358,15 +393,37 @@ pub mod pallet {
...
+						log::debug!("Detached hash: {:?} signature: {:?}", detach_hash, data.remote_signature);
+					});
+
+				}
+			}
+
+			Self::deposit_event(Event::CollectionDetached {
+				merkle_root: data.merkle_root,
+				remote_signature: data.remote_signature,
+				collection_type: data.collection.get_type(),
```
Comment (sinkingsugar) [inline pallets/detach/src/lib.rs:434]: enum or nothing here.

File: pallets/detach/src/lib.rs (hunk 5 above)
Comment (sinkingsugar) [reply pallets/detach/src/lib.rs]: remove comment btw

File: pallets/detach/src/lib.rs [hunk 10]
```diff
@@ -575,14 +632,23 @@ pub mod pallet {
 			}
 		}
 
+		fn get_merkle_root(detach_collection: &DetachCollection) -> Hash256 {
+			let detach_hashes = detach_collection.get_abi_encoded_hashes();
+			// TODO Review - Should we sort the leaves?
+			// detach_hashes.sort_by(|a, b| Keccak256::hash(a).cmp(&Keccak256::hash(b)));
```
Comment (sinkingsugar) [inline pallets/detach/src/lib.rs]: can you check actual literature? it might be good?

File: pallets/detach/src/lib.rs [hunk 11]
```diff
@@ -610,7 +679,8 @@ pub mod pallet {
 					}.to_big_endian(&mut chain_id_be);
 
 					let payload = [
-						&request.hash.get_signable_hash()[..],
+						&request.collection.get_type()[..],
```
Comment (sinkingsugar) [inline pallets/detach/src/lib.rs]: use an enum

**Review by sinkingsugar — CHANGES_REQUESTED** [review]

File: pallets/detach/src/lib.rs (hunk 10 above)
Comment (aang114) [reply pallets/detach/src/lib.rs]: It makes sense since then the **user won't have to care about the order of the leaves** (although btw in the emitted event `CollectionDetached` we show the order). 

So it's not actually necessary or better, but can be more convenient for the user. **I guess I'll just implement it then...**

File: pallets/detach/src/lib.rs (hunk 9 above)
Comment (aang114) [reply pallets/detach/src/lib.rs:434]: done

File: pallets/detach/src/lib.rs (hunk 11 above)
Comment (aang114) [reply pallets/detach/src/lib.rs]: done

**Comment by aang114** [comment]
Hi @sinkingsugar, please approve this PR (if okay) so that it can be merged. Thanks

File: pallets/detach/src/lib.rs [hunk 12]
```diff
@@ -633,9 +644,9 @@ pub mod pallet {
 		}
 
 		fn get_merkle_root(detach_collection: &DetachCollection) -> Hash256 {
-			let detach_hashes = detach_collection.get_abi_encoded_hashes();
+			let mut detach_hashes = detach_collection.get_abi_encoded_hashes();
 			// TODO Review - Should we sort the leaves?
```
Comment (sinkingsugar) [inline pallets/detach/src/lib.rs]: remove?

File: pallets/detach/src/lib.rs [hunk 13]
```diff
@@ -679,7 +690,10 @@ pub mod pallet {
 					}.to_big_endian(&mut chain_id_be);
 
 					let payload = [
-						&request.collection.get_type()[..],
+						// TODO Review - In `CollectionFactory.sol` - should we have converted the enum `CollectionType` to `uint8` or `uint64`/`uint256`?
+						// In Fragnova's Smart Contract `CollectionFactory.sol`, we convert the enum `CollectionType` to `uint8` when verifying the signature.
+						// Therefore, when constructing the signature - we are also converting it to `u8`.
+						&(request.collection.get_type() as u8).to_be_bytes()[..],
```
Comment (sinkingsugar) [inline pallets/detach/src/lib.rs:694]: uint8 is fine

**Review by sinkingsugar — CHANGES_REQUESTED** [review]

**Review by sinkingsugar — APPROVED** [review]
